src/sentry/hybridcloud/models/outbox.py

```diff
@@ -219,11 +220,143 @@ def process_shard(self, latest_shard_row: OutboxBase | None) -> Generator[Outbox
+        """
+        using = router.db_for_write(type(self))
+        tags: dict[str, int | str] = {"category": "None", "synchronous": int(is_synchronous_flush)}
+
+        # If we have a latest_row, only operate on messages that were created before it.
+        filters = {} if latest_shard_row is None else dict(id__lte=latest_shard_row.id)
+        try:
+            now = timezone.now()
+            with transaction.atomic(using=using, savepoint=False):
+                # Lock all the rows we're going to process and attempt to reserve them.
+                reserved_ids = list(
+                    self.select_coalesced_messages()
+                    .filter(scheduled_for__lte=now, **filters)
+                    .select_for_update(nowait=True)
+                    .order_by("id")
+                    .values_list("id", flat=True)
+                )
+                if not reserved_ids:
+                    yield None
+                    return
+
+                coalesced: OutboxBase = self.objects.get(id=reserved_ids[-1])
+                first_coalesced: OutboxBase = (
+                    self.objects.filter(id=reserved_ids[0]).first() or coalesced
+                )
+
+                # Reserve all messages in the coalesce group so they cannot
+                # be selected by concurrent calls.
+                self.objects.filter(
+                    id__in=reserved_ids,
+                ).update(scheduled_for=coalesced.next_schedule(now))
+        except OperationalError as e:
+            # If concurrent locking is happening on the table, gracefully pass and allow
+            # that work to process.
+            if "LockNotAvailable" in str(e):
+                yield None
+                return
+            else:
+                raise
+
+        tags["category"] = OutboxCategory(self.category).name
+        assert first_coalesced, "first_coalesced incorrectly set for non-empty coalesce group"
+        metrics.timing(
+            "outbox.coalesced_net_queue_time",
+            now.timestamp() - first_coalesced.date_added.timestamp(),
+            tags=tags,
+        )
+
+        # Yield the latest message for processing outside of a transaction
+        yield coalesced
+
+        # If the block didn't raise, delivery was successful, and we
+        # can delete messages to acknowledge them.
+        try:
+            deleted_count = 0
+            with transaction.atomic(using=using):
+                # The last message in a coalesce group should be preserved if we're
+                # skipping the shard, as we need to delay signals, and keep
+                # backlogs shorter, but not lose updates.
+                cleanup_ids = reserved_ids[:-1]
+                if cleanup_ids:
+                    self.objects.filter(id__in=cleanup_ids).delete()
+                    deleted_count = len(cleanup_ids)
+                if not self.should_skip_shard():
+                    deleted_count += 1
+                    coalesced.delete()
```

Comment:
If a shard is skipped we retain the 'latest' record but remove all other records in the coalesce group. This condition had a test and I preserved behavior in the new implementation.